crates/sui-indexer/src/indexer_reader.rs

```diff
@@ -287,61 +264,98 @@ impl IndexerReader {
         Ok(Some(epoch_info))
     }
 
-    fn get_epochs_from_db(
+    async fn get_epochs_from_db(
         &self,
         cursor: Option<u64>,
         limit: usize,
         descending_order: bool,
     ) -> Result<Vec<StoredEpochInfo>, IndexerError> {
-        run_query!(&self.pool, |conn| {
-            let mut boxed_query = epochs::table.into_boxed();
-            if let Some(cursor) = cursor {
-                if descending_order {
-                    boxed_query = boxed_query.filter(epochs::epoch.lt(cursor as i64));
-                } else {
-                    boxed_query = boxed_query.filter(epochs::epoch.gt(cursor as i64));
-                }
-            }
+        use diesel_async::RunQueryDsl;
+
+        let mut connection = self.pool.get().await?;
+
+        let mut query = epochs::table.into_boxed();
+
+        if let Some(cursor) = cursor {
             if descending_order {
-                boxed_query = boxed_query.order_by(epochs::epoch.desc());
+                query = query.filter(epochs::epoch.lt(cursor as i64));
             } else {
-                boxed_query = boxed_query.order_by(epochs::epoch.asc());
+                query = query.filter(epochs::epoch.gt(cursor as i64));
             }
+        }
 
-            boxed_query.limit(limit as i64).load(conn)
-        })
+        if descending_order {
+            query = query.order_by(epochs::epoch.desc());
+        } else {
+            query = query.order_by(epochs::epoch.asc());
+        }
+
+        query
+            .limit(limit as i64)
+            .load(&mut connection)
+            .await
+            .map_err(Into::into)
     }
 
-    pub fn get_epochs(
+    pub async fn get_epochs(
         &self,
         cursor: Option<u64>,
         limit: usize,
         descending_order: bool,
     ) -> Result<Vec<EpochInfo>, IndexerError> {
-        self.get_epochs_from_db(cursor, limit, descending_order)?
+        self.get_epochs_from_db(cursor, limit, descending_order)
+            .await?
             .into_iter()
             .map(EpochInfo::try_from)
             .collect::<Result<Vec<_>, _>>()
             .map_err(Into::into)
     }
 
-    pub fn get_latest_sui_system_state(&self) -> Result<SuiSystemStateSummary, IndexerError> {
-        let system_state: SuiSystemStateSummary =
-            sui_types::sui_system_state::get_sui_system_state(self)?
-                .into_sui_system_state_summary();
+    pub async fn get_latest_sui_system_state(&self) -> Result<SuiSystemStateSummary, IndexerError> {
+        let object_store = ConnectionAsObjectStore::from_pool(&self.pool)
+            .await
+            .map_err(|e| IndexerError::PgPoolConnectionError(e.to_string()))?;
+
+        let system_state = tokio::task::spawn_blocking(move || {
+            sui_types::sui_system_state::get_sui_system_state(&object_store)
```

Comment:
can we use self.get_pool (the blocking pool) here?

Reply:
no, we're trying to eliminate the blocking pool entirely